### ISSUE TITLE
RabbitMQ's Direct reply-to

### ISSUE THREAD

**Issue by anton-shirikov** [pdezwart/php-amqp#155]
I've been trying to implement RabbitMQ's [direct reply-to](https://www.rabbitmq.com/direct-reply-to.html) and there seems to be a problem.

#### First attempt:
```php
// Establish connection, create a channel & exchange for publishing...
$replyQueue = new AMQPQueue($channel);
$replyQueue->setName('amq.rabbitmq.reply-to');

echo 'Publishing...' . PHP_EOL;
$exchange->publish('ping', null, AMQP_NOPARAM, [
    'reply_to' => 'amq.rabbitmq.reply-to',
]);

echo 'Waiting for reply...' . PHP_EOL;
$replyQueue->consume(function (AMQPEnvelope $envelope) {
   echo $envelope->getBody() . PHP_EOL;
}, AMQP_AUTOACK);
```

With amqp 1.4.0 the producer hangs on consume and rabbitmq logs contain this error:
```
Channel error on connection <0.601.4> (172.17.0.105:48186 -> 172.17.0.103:5672, vhost: '/', user: 'admin'), channel 1:
{amqp_error,precondition_failed,"fast reply consumer does not exist",
            'basic.publish'}
```
With amqp 1.6.0beta2 producer recognizes the error and throws exception:
```
PHP Fatal error:  Uncaught exception 'AMQPQueueException' with message 'Server channel error: 406, message: PRECONDITION_FAILED - fast reply consumer does not exist' in /var/code/producer.php:26
```

#### Second attempt (kinda working):

Judging from this reply http://permalink.gmane.org/gmane.comp.networking.rabbitmq.general/36792
I need to first consume from `amq.rabbitmq.reply-to`  before i send a request.

I've tried to find a workaround with amqp-1.6.0beta2's consume from multiple queues and it is working. The problem is I have to declare an extra "useless" queue:

```php
// Establish connection, create a channel & exchange for publishing...
$queue = new AMQPQueue($channel);
$queue->setName('amq.rabbitmq.reply-to');
$queue->consume(null, AMQP_AUTOACK);

$queue2 = new AMQPQueue($channel);
$queue2->setName('temp-queue');
$queue2->declareQueue();

echo 'Publishing...' . PHP_EOL;
$exchange->publish('ping', null, AMQP_NOPARAM, [
    'reply_to' => 'amq.rabbitmq.reply-to',
]);

echo 'Waiting for reply...' . PHP_EOL;
$queue2->consume(function (AMQPEnvelope $envelope) {
   echo $envelope->getBody() . PHP_EOL;
}, AMQP_AUTOACK);

$queue->cancel();
$queue2->cancel();
```

#### php-amqplib

With php-amqplib library there seems to be no problem and everything is working as expected:

```php
// Establish connection, create a channel...
$callback_queue = 'amq.rabbitmq.reply-to';

$response = null;
$channel->basic_consume($callback_queue, '', false, true, false, false, function (AMQPMessage $message) use (&$response) {
    echo $message->body . PHP_EOL;
    $response = true;
});

$msg = new AMQPMessage('ping', [
    'reply_to' => $callback_queue
]);

$channel->basic_publish($msg, 'exchange_name');
while (!$response) {
    $channel->wait();
}
```

#### Consumer
Consumer seems to be irrelevant. I've used the same script for all 3 cases:

```php
$exchange = new AMQPExchange($channel);
$exchange->setName('exchange_name');
$exchange->setType(AMQP_EX_TYPE_DIRECT);
$exchange->declareExchange();

$queue = new AMQPQueue($channel);
$queue->setName('queue');
$queue->setFlags(AMQP_NOPARAM);
$queue->declareQueue();
$queue->bind($exchange->getName());

$defaultExchange = new AMQPExchange($channel);

echo "Waiting for messages..." . PHP_EOL;
$queue->consume(function (AMQPEnvelope $envelope) use ($queue, $defaultExchange) {
    printf("Got message `%s` with reply-to `%s`" . PHP_EOL, $envelope->getBody(), $envelope->getReplyTo());
    $defaultExchange->publish('pong', $envelope->getReplyTo());
}, AMQP_AUTOACK);
```

**Comment by anton-shirikov** [comment]
Any feedback ? Is there any chance this will be implemented in the near future?

Thank you

**Comment by pinepain** [comment]
Hi Anton, sorry for late feedback. I am a bit overcommited last time.

With the following script:

```php
<?php
$conn = new AMQPConnection();
$conn->connect();

$channel = new AMQPChannel($conn);

$replyQueue = new AMQPQueue($channel);
$replyQueue->setName('amq.rabbitmq.reply-to');
$replyQueue->consume(null, AMQP_AUTOACK);


$exchange = new AMQPExchange($channel);

echo 'Publishing...' . PHP_EOL;
$exchange->publish('ping', null, AMQP_NOPARAM, [
    'reply_to' => 'amq.rabbitmq.reply-to',
]);

echo 'Waiting for reply...' . PHP_EOL;
$replyQueue->consume(function (AMQPEnvelope $envelope) {
   echo $envelope->getBody() . PHP_EOL;
}, AMQP_AUTOACK);
``` 
inside projects Vagrant box I get such output:

```
vagrant@vagrant-ubuntu-utopic-64 (php:5.6.4 -phpbrew):~/php-amqp$ php test.php 
Publishing...
Waiting for reply...
PHP Fatal error:  Uncaught exception 'AMQPQueueException' with message 'Server channel error: 406, message: PRECONDITION_FAILED - reply consumer already set' in /home/vagrant/php-amqp/test.php:22
Stack trace:
#0 /home/vagrant/php-amqp/test.php(22): AMQPQueue->consume(Object(Closure), 128)
#1 {main}
  thrown in /home/vagrant/php-amqp/test.php on line 22

Fatal error: Uncaught exception 'AMQPQueueException' with message 'Server channel error: 406, message: PRECONDITION_FAILED - reply consumer already set' in /home/vagrant/php-amqp/test.php on line 22

AMQPQueueException: Server channel error: 406, message: PRECONDITION_FAILED - reply consumer already set in /home/vagrant/php-amqp/test.php on line 22

Call Stack:
    0.0005     243712   1. {main}() /home/vagrant/php-amqp/test.php:0
    0.0062     250904   2. AMQPQueue->consume() /home/vagrant/php-amqp/test.php:22
```

which is definitely not what we want.

I will work on a bug fix within a few days.

I guess we shouldn't send multiple `basic.consume` requests to the same queue, one is enough (and most time this is desired behavior). Keeping in mind that this extension based on pretty legacy codebase (it is not so bad, just with large history) storing in `HashTable `queues name as keys and consumer tags as values may serve as a a temporary solution.

**Comment by pinepain** [comment]
Can you attach minimal examples that works for you (both php-amqp and phpamqplib)?

**Comment by anton-shirikov** [comment]
Thank you for your reply.

I suppose we need a non-blocking ```basic.consume``` and a way to consume later without sending ```basic.consume``` again (Exactly what you pointed out). 

> Can you attach minimal examples that works for you (both php-amqp and phpamqplib)?

phpamqplib in my first post example is working
Conceptually your example for php-amqp is valid as well

**Comment by pinepain** [comment]
Sorry, I can't get it working. I even post a request (https://groups.google.com/forum/#!topic/rabbitmq-users/wHYwE6DKuKE) to [rabbitmq-users group](https://groups.google.com/forum/#!topic/rabbitmq-users).

Which RabbitMQ version do you run your example on? Can you post complete code to reproduce desired behavior with php-amqplib?

**Comment by anton-shirikov** [comment]
RabbitMQ version is 3.5.0.
I've used docker to setup environment. If it helps, I can post Dockerfile / docker-compose.yml config that was used

Code:

producer.php
```php
<?php

use PhpAmqpLib\Connection\AMQPConnection;
use PhpAmqpLib\Message\AMQPMessage;

require_once __DIR__ . '/../vendor/autoload.php';

$connection = new AMQPConnection(
    'rabbitmq', 5672, 'admin', getenv('RABBITMQ_ENV_RABBITMQ_PASS'));

$channel = $connection->channel();

$callback_queue = 'amq.rabbitmq.reply-to';

$response = null;
$channel->basic_consume($callback_queue, '', false, true, false, false, function (AMQPMessage $message) use (&$response) {
    echo $message->body . PHP_EOL;
    $response = true;
});

$msg = new AMQPMessage('ping', [
    'reply_to' => $callback_queue
]);

$channel->basic_publish($msg, 'exchange_name');
while (!$response) {
    $channel->wait();
}
```

consumer.php writtern using php-amqp extension (i've used same consumer as for my php-amqp tests)
```php
<?php

$connection = new AMQPConnection([
    'host' => 'rabbitmq',
    'login' => 'admin',
    'password' => getenv('RABBITMQ_ENV_RABBITMQ_PASS')
]);

$connection->connect();
$channel = new AMQPChannel($connection);
$channel->setPrefetchCount(1);

$exchange = new AMQPExchange($channel);
$exchange->setName('message_exchange');
$exchange->setType(AMQP_EX_TYPE_DIRECT);
$exchange->declareExchange();

$queue = new AMQPQueue($channel);
$queue->setName('ttl-expires-queue');
$queue->setFlags(AMQP_NOPARAM | AMQP_DURABLE);
$queue->setArgument('x-dead-letter-exchange', 'dlx-exchange');
$queue->declareQueue();
$queue->bind($exchange->getName());

$defaultExchange = new AMQPExchange($channel);

echo "Waiting for messages..." . PHP_EOL;
$queue->consume(function (AMQPEnvelope $envelope) use ($queue, $defaultExchange) {
    printf('Got message #`%s` with body `%s`' . PHP_EOL, $envelope->getDeliveryTag(), $envelope->getBody());
    $defaultExchange->publish($envelope->getBody(), $queue->getName(), AMQP_NOPARAM, $envelope->getHeaders());
    $queue->ack($envelope->getDeliveryTag());
    sleep(1);
//    $queue->ack($envelope->getDeliveryTag());
}, AMQP_NOPARAM);

```

**Comment by pinepain** [comment]
Thanks. Firs time I was biased a bit by partial examples you provided (I wasn't familiar with direct reply-to right before yesterday night). I made changes that now allow us to use direct reply-to, see PR https://github.com/pdezwart/php-amqp/pull/158.

**Comment by anton-shirikov** [comment]
Thank you

**Comment by pinepain** [comment]
You're welcome. Check the latest version now, there was dumb mistakes which should be fixed now (tests are green).

**Comment by anton-shirikov** [comment]
Well it looks valid to me. I am not sure when I'll get the time to verify it myself, sorry.

**Comment by pinepain** [comment]
@lstrojny this can be closed while fix merget in #158
